easynav_gps_vff/easynav_gps_vff_controller/src/easynav_gps_vff_controller/GpsVffController.cpp

```diff
@@ -0,0 +1,306 @@
+  }
+
+  // Combine attractive and repulsive vectors
+  vff_vector.result[0] = vff_vector.attractive[0] + vff_vector.repulsive[0];
+  vff_vector.result[1] = vff_vector.attractive[1] + vff_vector.repulsive[1];
+
+  // Publish debug markers
+  visualization_msgs::msg::Marker marker;
+  marker.header.frame_id = cmd_vel_.header.frame_id;
+  marker.header.stamp = get_node()->now();
+  marker.type = visualization_msgs::msg::Marker::SPHERE;
+  marker.id = 456;
+  marker.pose.position.x = closest_point.x;
+  marker.pose.position.y = closest_point.y;
+  marker.pose.position.z = closest_point.z;
+  marker.pose.orientation.w = 1.0;
+  marker.scale.x = marker.scale.y = marker.scale.z = 0.1;
+  marker.color.a = 1.0;
+  marker.color.r = 1.0;
+  marker.color.g = 0.0;
+  marker.color.b = 0.0;
+
+  auto marker_array = get_debug_vff(vff_vector, frame_id);
+  marker_array.markers.push_back(marker);
+
+  if (marker_array_pub_->get_subscription_count() > 0) {
+    marker_array_pub_->publish(marker_array);
+  }
+
+  return vff_vector;
+}
+
+void GpsVffController::update_rt(const NavState & nav_state)
+{
+  nav_msgs::msg::Goals all_goals = nav_state.goals;
+  nav_msgs::msg::Odometry odom = nav_state.odom;
+
+  // Current position
+  double current_x_ = odom.pose.pose.position.x;
+  double current_y_ = odom.pose.pose.position.y;
+
+  // If a goal is set
+  if (!all_goals.goals.empty()) {
+    // Nuevo goal recibido
+    geometry_msgs::msg::Point new_goal;
+    new_goal.x = all_goals.goals[0].pose.position.x;
+    new_goal.y = all_goals.goals[0].pose.position.y;
+
+    // Comprobar si es diferente al anterior
+    bool new_goal_received = !has_previous_goal_ ||
+      std::hypot(new_goal.x - previous_goal_.x, new_goal.y - previous_goal_.y) > 0.01;
+
+    if (new_goal_received) {
+      RCLCPP_INFO(get_node()->get_logger(), "New goal [%f, %f]", new_goal.x, new_goal.y);
+      target_reached_ = false;
+      previous_goal_ = new_goal;
+      has_previous_goal_ = true;
+    }
+
+    goal_ = new_goal;
+
+    // Calculate the difference in position
+    double dx = goal_.x - current_x_;
+    double dy = goal_.y - current_y_;
+
+    // Calculate the Euclidean distance to the goal
+    double distance = std::hypot(dx, dy);  // more stable than sqrt(pow(..., 2))
+
+    // Calculate the angle to the goal
+    double bearing = normalize_angle(std::atan2(dy, dx));
+
+    // Calculate the angle error
+    double angle_error = normalize_angle(bearing - yaw_);
+
+    if (distance < distance_to_goal_) {
+      if (target_reached_) {return;}
+      // Target reached → stop the robot
+      RCLCPP_INFO(get_node()->get_logger(), "Target reached");
+      cmd_vel_.twist.linear.x = 0.0;
+      cmd_vel_.twist.angular.z = 0.0;
+      goal_ = geometry_msgs::msg::Point();
+      target_reached_ = true;
+      return;
+    } else {
+      target_reached_ = false;
+    }
+
+    // Get the fused PointCloud
+    auto fused =
```

Comment:
We should take time measurements to ensure that this method takes less than ~10ms. 

I would start this pipeline with a filter to consider only the nearby points _a grosso modo_, and check it if this makes this line more efficient, as the `fuse` is the expensive step in this pipeline :

```
auto fused =
      PerceptionsOpsView(nav_state.perceptions)
            .filter({-10.0, -10.0, -10.0}, ({10.0, 10.0, 10.0})
            .fuse(cmd_vel_.header.frame_id)
            ->filter({obstacle_detection_x_min_, obstacle_detection_y_min_, obstacle_detection_z_min_{obstacle_detection_x_max_, obstacle_detection_y_max_,
        obstacle_detection_z_max_}).as_points();
```